shell/platform/darwin/common/availability_version_check.cc

```diff
@@ -2,20 +2,129 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+#include "flutter/shell/platform/darwin/common/availability_version_check.h"
+
+#include <cstdint>
+#include <optional>
+
+#include <CoreFoundation/CoreFoundation.h>
 #include <dispatch/dispatch.h>
 #include <dlfcn.h>
-#include <cstdint>
 
+#include "flutter/fml/build_config.h"
+#include "flutter/fml/file.h"
 #include "flutter/fml/logging.h"
+#include "flutter/fml/mapping.h"
+#include "flutter/fml/platform/darwin/cf_utils.h"
+
+// The implementation of _availability_version_check defined in this file is
+// based on the code in the clang-rt library at:
+//
+// https://github.com/llvm/llvm-project/blob/e315bf25a843582de39257e1345408a10dc08224/compiler-rt/lib/builtins/os_version_check.c
+//
+// Flutter provides its own implementation due to an issue introduced in recent
+// versions of Clang following Clang 18 in which the clang-rt library declares
+// weak linkage against the _availability_version_check symbol. This declaration
+// causes apps to be rejected from the App Store. When Flutter statically links
+// the implementation below, the weak linkage is satisfied at Engine build time,
+// the symbol is no longer exposed from the Engine dylib, and apps will then
+// not be rejected from the App Store.
+//
+// The implementation of _availability_version_check can delegate to the
+// dynamically looked-up symbol on recent iOS versions, but the lookup will fail
+// on iOS 11 and 12. When the lookup fails, the current OS version must be
+// retrieved from a plist file at a well-known path. The logic for this below is
+// copied from the clang-rt implementation and adapted for the Engine.
 
-// See context in https://github.com/flutter/flutter/issues/132130 and
+// See more context in https://github.com/flutter/flutter/issues/132130 and
 // https://github.com/flutter/engine/pull/44711.
 
 // TODO(zanderso): Remove this after Clang 18 rolls into Xcode.
-// https://github.com/flutter/flutter/issues/133203
+// https://github.com/flutter/flutter/issues/133203.
+
+#define CF_PROPERTY_LIST_IMMUTABLE 0
+
+namespace flutter {
+
+// This function parses the platform's version information out of a plist file
+// at a well-known path. It parses the plist file using CoreFoundation functions
+// to match the implementation in the clang-rt library.
+std::optional<ProductVersion> ProductVersionFromSystemVersionPList() {
+  std::string plist_path = "/System/Library/CoreServices/SystemVersion.plist";
```

Comment:
I'll take your word for these paths and env variables :)

Reply:
🤞🏻